### PR TITLE
Fix concurrent transactions through removal of nested transactions

### PR DESCRIPTION
**Pull request by thehabbos007**
## 🌟 What is the purpose of this PR?

This PR moves away from allowing nested transactions in our PostgreSQL database layer. It has been found that subtransactions are not ideal for hot code paths as seen [here](https://postgres.ai/blog/20210831-postgresql-subtransactions-considered-harmful) and [here](https://about.gitlab.com/blog/2021/09/29/why-we-spent-the-last-month-eliminating-postgresql-subtransactions/). 

We've also seen issues with these nested transactions in the DB library we use: slonik. 
Since we would like to concurrently await for lists of promises whenever possible, we run into `Cannot run a query using parent transaction` when using `Promise.all` on a list of promises that go to the database (and use transactions). This happens because slonik tracks transaction nesting sequentially.
https://github.com/gajus/slonik/blob/c6ddd79a78518a9dd437faa7196424368c245b27/src/binders/bindTransactionConnection.ts#L38-L42

On top of that, subtransactions are assigned a unique transaction id of which only 64 can be cached at a time https://github.com/postgres/postgres/blob/78ab944cd4b9977732becd9d0bc83223b88af9a2/src/include/storage/proc.h#L36
This limitation might become a performance issue at some point.

These factors together prove subtransactions as a restriction that this PR fixes.

The Database transaction connection type has now been changed such that the `transaction` function is omitted
and the different connections are tagged such that we can check the kind of connection we're dealing with at runtime
https://github.com/hashintel/hash/blob/dab09b221abebd777f501652aad430f69c52bae4/packages/hash/api/src/db/postgres/types.ts#L6-L13

When a transaction is requested (through the below function), any top-level connection will create a connection, and any transaction connection will skip creating a transaction
https://github.com/hashintel/hash/blob/dab09b221abebd777f501652aad430f69c52bae4/packages/hash/api/src/db/postgres/util.ts#L20-L30

This means that we can only have a single transaction level, while still allowing transactions anywhere. The fix here is a drop-in replacement for anywhere we were calling `connection.transaction`.
Resolvers can now create a transaction, flattening every subsequent transaction that, for example, model classes or DB adapter methods require. 
Alternatively resolvers can call down to model classes, deferring creation of a transaction to the point where it is needed.

## 🚫 Blocked by

- [ ] ...

## 🐾 Next steps

- Re-enable concurrent transactions anywhere it has been missed

## 🔍 What does this change?

- Wrap `Connection` type of `slonik` with some restrictions and allow for runtime type matching
  - Wrapping happens through new constructor functions in `types.ts` for postgres
- Flatten transactions through unifying transactions in function that used postgres transactions
  - `client.ts` and various files dealing with SQL directly
- Add tests for the new transaction flattening
- Revert to concurrency at known locations where concurrency were used before

Note that because of the type changes, it is not possible to pass type checking if one were to try to call `connection.transaction` without going through the helper method. 

### 📜 Does this require a change to the docs?

- The docs for x need updating to explain that y

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1201095311341924/1201817757879404/f) (_internal_)

## 🛡 What tests cover this?

- Unit tests
- Existing integration tests
- Some manual testing for collab

## ❓ How to test this?

This doesn't change existing functionality, it should just be a performance boost. Benchmarks on their way

## 📹 Demo (performance)

Inserting 21 blocks into a blank page
1 divider
20 text blocks

**Concurrently**: ~800ms
**Sequentially**: ~950ms

So the concurrent transactions do make a difference. And the gains would compound as we deal with more DB queries that can be run concurrently..

### CODE REVIEW

File: packages/hash/api/src/db/postgres/types.ts [hunk 1]
```diff
@@ -3,6 +3,54 @@ import {
   DatabaseTransactionConnectionType,
 } from "slonik";
 
+// The variant kinds are distinct types.
+// Enum and union types unify at the top level type, which isn't desired for these typings.
+export type PoolConnectionKind = "PoolConnectionKind";
+export type TransactionConnectionKind = "TransactionConnection";
```
Comment (thehabbos007) [inline packages/hash/api/src/db/postgres/types.ts:9]: I am not too happy with having to do this. Unfortunately both union types and enums would behave weirdly in the tagged sum type, which would try to unify types in a strange way. If anyone has a good way to allow runtime tags, please do tell!

File: packages/hash/api/src/db/postgres/util.test.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
+import { transaction } from "./util";
+import { ConnectionKind, PoolConnection, TransactionConnection } from "./types";
+
+describe("transaction flattening", () => {
```
Comment (thehabbos007) [inline packages/hash/api/src/db/postgres/util.test.ts:4]: Ideally we could also check that the SQL itself doesn't contain any savepoints, but this already is quite coupled to slonik.

File: packages/hash/api/src/db/postgres/types.ts [hunk 3]
```diff
@@ -3,6 +3,37 @@ import {
...
+ * Postgres connection types.
+ * Note that {@link TransactionConnection} does not allow for nested transactions.
+ * Construction of these types can be done through {@link createPoolConnection} or {@link createTransactionConnection}
+ */
+export type Connection = PoolConnection | TransactionConnection;
+
+/**
+ * Construct {@link PoolConnection} variant tagged with a type
+ * @param database pool connection
+ * @returns a {@link Connection}
```
Comment (benwerner01) [inline packages/hash/api/src/db/postgres/types.ts:25]:
```suggestion
 * @returns a {@link PoolConnection}
```

File: packages/hash/api/src/db/postgres/types.ts [hunk 4]
```diff
@@ -3,6 +3,37 @@ import {
...
+ */
+export const createPoolConnection = (
+  connection: DatabasePoolConnectionType,
+): PoolConnection => Object.assign(connection, { type: "Pool" } as const);
+
+/**
+ * Construct {@link TransactionConnection} variant tagged a type
+ * @param database transaction connection
+ * @returns a {@link Connection}
```
Comment (benwerner01) [inline packages/hash/api/src/db/postgres/types.ts:34]:
```suggestion
 * @returns a {@link TransactionConnection}
```

File: packages/hash/api/src/db/postgres/util.test.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
+import { transaction } from "./util";
+import { PoolConnection, TransactionConnection } from "./types";
+
+describe("transaction flattening", () => {
+  it("allows a single top-level transaction", async () => {
+    const transactionsHandler = jest.fn().mockResolvedValue("ok");
+    const transactionConn = {
+      // The transaction connection is untagged when it is created by the DB
+      // the type field will be present after creating a transaction.
+      ...({} as TransactionConnection),
```
Comment (benwerner01) [inline packages/hash/api/src/db/postgres/util.test.ts:10]:
```suggestion
      ...({ type: "Transaction" } as TransactionConnection),
```

**Review by benwerner01 — APPROVED** [review]
Thanks @thehabbos007 , looks great 👍

Made a few minor suggestions for renaming some things, but nothing that blocks merging :)

File: packages/hash/api/src/db/postgres/types.ts [hunk 6]
```diff
@@ -3,6 +3,54 @@ import {
...
+export type TransactionConnectionKind = "TransactionConnection";
+
+/**
+ * The types given for {@link PoolConnectionKind} and {@link TransactionConnectionKind} cannot be used as values directly.
+ * This object maps a connection kind to its type.
+ */
+export const ConnectionKind = {
+  PoolConnection: <PoolConnectionKind>"PoolConnectionKind",
+  TransactionConnection: <TransactionConnectionKind>"TransactionConnectionKind",
+};
```
[the remark this answers is not in the document]
Comment (benwerner01) [reply packages/hash/api/src/db/postgres/types.ts]: Outdated observation, resolving

File: packages/hash/api/src/db/postgres/util.test.ts (hunk 5 above)
Comment (thehabbos007) [reply packages/hash/api/src/db/postgres/util.test.ts:10]: This is left out intentionally to check if the connection is properly tagged when a transaction is requested from the poolconneciton

File: packages/hash/api/src/db/postgres/util.ts [hunk 7]
```diff
@@ -1,4 +1,26 @@
...
+export { poolConnection, transactionConnection } from "./types";
+
+/**
+ * Conditionally create transaction or use active transaction depending on connection kind.
+ * If the connection given is a top-level connection i.e. not a transaction, a transaction will be created.
+ * Otherwise we simply make use of the transaction connection as is.
+ * @param connection either top-level or one transaction level.
+ * @returns handler callback
+ */
+export const transaction =
```
[the remark this answers is not in the document]
Comment (thehabbos007) [reply packages/hash/api/src/db/postgres/util.ts:20]: That's a good distinction. I will apply this

File: packages/hash/api/src/db/postgres/util.test.ts (hunk 5 above)
Comment (benwerner01) [reply packages/hash/api/src/db/postgres/util.test.ts:10]: Ah apologies, the type confused me but I see what's going on now.

**Review by benwerner01 — APPROVED** [review]
Thanks! 👌

**Review by benwerner01 — APPROVED** [review]

**Review by CiaranMn — APPROVED** [review]
Neat!

**Pull request merged**